game/src/edit/traffic_signals/mod.rs

```diff
@@ -421,18 +463,101 @@ impl State for NewTrafficSignalEditor {
     }
 }
 
+fn make_top_panel(ctx: &mut EventCtx, app: &App, can_undo: bool, can_redo: bool) -> Composite {
+    let row = vec![
+        Btn::text_fg("Finish").build_def(ctx, hotkey(Key::Escape)),
+        Btn::text_fg("Preview").build_def(ctx, lctrl(Key::P)),
+        (if can_undo {
+            Btn::svg_def("system/assets/tools/undo.svg").build(ctx, "undo", lctrl(Key::Z))
+        } else {
+            Widget::draw_svg_transform(
+                ctx,
+                "system/assets/tools/undo.svg",
+                RewriteColor::ChangeAll(Color::WHITE.alpha(0.5)),
+            )
+        })
+        .centered_vert(),
+        (if can_redo {
+            Btn::svg_def("system/assets/tools/redo.svg").build(
+                ctx,
+                "redo",
+                // TODO ctrl+shift+Z!
+                lctrl(Key::Y),
+            )
+        } else {
+            Widget::draw_svg_transform(
+                ctx,
+                "system/assets/tools/redo.svg",
+                RewriteColor::ChangeAll(Color::WHITE.alpha(0.5)),
+            )
+        })
+        .centered_vert(),
+        if app.opts.dev {
+            Btn::text_fg("Export")
+                .tooltip(Text::from_multiline(vec![
+                    Line("This will create a JSON file in traffic_signal_data/.").small(),
+                    Line(
+                        "Contribute this to map how this traffic signal is currently timed in \
+                         real life.",
+                    )
+                    .small(),
+                ]))
+                .build_def(ctx, None)
+        } else {
+            Widget::nothing()
+        },
+    ];
+    Composite::new(Widget::col(vec![
+        Line("Traffic signal editor").small_heading().draw(ctx),
+        Widget::row(row),
+    ]))
+    .aligned(HorizontalAlignment::Center, VerticalAlignment::Top)
+    .build(ctx)
+}
+
 fn make_side_panel(
     ctx: &mut EventCtx,
     app: &App,
     members: &BTreeSet<IntersectionID>,
     selected: usize,
 ) -> Composite {
     let map = &app.primary.map;
-
-    let mut col = Vec::new();
-
     // Use any member for phase duration
     let canonical_signal = map.get_traffic_signal(*members.iter().next().unwrap());
+
+    let mut txt = Text::new();
+    if members.len() == 1 {
+        let i = *members.iter().next().unwrap();
+        txt.add(Line(i.to_string()).big_heading_plain());
+
+        let mut road_names = BTreeSet::new();
+        for r in &app.primary.map.get_i(i).roads {
+            road_names.insert(app.primary.map.get_r(*r).get_name());
+        }
+        for r in road_names {
+            txt.add(Line(format!("- {}", r)));
+        }
+    } else {
+        txt.add(Line(format!("{} intersections", members.len())).big_heading_plain());
+    }
+    {
+        let mut total = Duration::ZERO;
+        for p in &canonical_signal.phases {
+            total += p.phase_type.simple_duration();
+        }
+        // TODO Say "normally" to account for adaptive phases?
```

Comment:
hmmm just thinking about adaptive phases... 

If I select 4 signals to edit, and give them all adaptive phasing

Initially, they'll switch in time with eachother, but if there's traffic waiting at just one of the lights - it'll adaptively make the phase longer. Won't that cause it to get out of sync with the rest of it's edit group?

I think that's probably not a small fix and a reasonable limitation for the current system, I just wanted to see if I understand how this currently works...

Reply:
Urgh, this is indeed a nasty case I hadn't anticipated. Originally I threw together the adaptive phase type as a proof-of-concept for https://github.com/dabreegster/abstreet/issues/91. The policy as it exists is pretty bad; a single car will repeat the entire phase for the full duration. I think there will have to be substantial future work to understand how actuated/adaptive signals work in reality, model that in abst, and keep the plethora of configuration options sane to understand in the UI.